crates/ra_hir/src/code_model_api.rs

```diff
@@ -562,7 +564,55 @@ impl Function {
     }
 
     pub fn diagnostics(&self, db: &impl HirDatabase, sink: &mut DiagnosticSink) {
-        self.infer(db).add_diagnostics(db, *self, sink);
+        let infer = self.infer(db);
+        infer.add_diagnostics(db, *self, sink);
+        self.validate_body(infer, db, sink)
+    }
+
+    fn validate_body(
+        &self,
+        infer: Arc<InferenceResult>,
+        db: &impl HirDatabase,
+        sink: &mut DiagnosticSink,
+    ) {
+        let file_id = self.source(db).0;
+        let body = self.body(db);
+        let source_map = self.body_source_map(db);
+        for e in body.exprs() {
+            match e {
+                (id, Expr::StructLit { path: _, fields, spread: None }) => {
+                    let lit_fields: HashMap<_, _> =
+                        fields.into_iter().map(|f| (&f.name, f)).collect();
+                    let def_fields = Function::struct_fields(&infer[id], db);
+                    let fills: Vec<StructLitFieldFill> = def_fields
+                        .into_iter()
+                        .map(|def_field| match lit_fields.get(&def_field) {
+                            Some(StructLitField { name, expr }) => StructLitFieldFill {
+                                name: name.to_string(),
+                                value: source_map.expr_syntax(*expr),
+                            },
+                            None => StructLitFieldFill { name: def_field.to_string(), value: None },
+                        })
+                        .collect();
+                    let unmatched = fills.iter().filter(|fill| fill.value == None).count();
+                    if unmatched > 0 {
+                        if let Some(ls) = source_map.field_list_syntax(id) {
+                            sink.push(FillStructFields { file: file_id, decl: ls, fills })
+                        } else {
+                            log::error!("No source mapping for {:#?} named fields", e);
```

Comment:
We generally just `.unwrap` in such cases. Probably even source_map should be adjusted to not return NOne